certbot/certbot/_internal/auth_handler.py

```diff
@@ -436,18 +407,29 @@ def _report_failed_authzrs(failed_authzrs, account_key):
     for achall in failed_achalls:
         problems.setdefault(achall.error.typ, []).append(achall)
 
-    reporter = zope.component.getUtility(interfaces.IReporter)
-    for achalls in problems.values():
-        reporter.add_message(_generate_failed_chall_msg(achalls), reporter.MEDIUM_PRIORITY)
+    config = zope.component.getUtility(interfaces.IConfig)
+    msg = ["\nCertbot failed to authenticate some domains (using the {} plugin)."
+           " The ACME server reported these problems:".format(config.authenticator)]
+
+    for _, achalls in sorted(problems.items(), key=lambda item: item[0]):
+        msg.append(_generate_failed_chall_msg(achalls))
+
+    msg.append(
```

Comment:
@ohemorange on Dec 8 we talked about whether it was possible to have some generic error text wording that wasn't overly presumptive (wrong) about what the problem was.

What do you think about:

>Certbot failed to authenticate some domains (using the webroot plugin). The ACME server reported these problems:
>  Domain: example.com
>  Type:   unauthorized
>  Detail: Invalid response from http://example.com/.well-known/acme-challenge/f0gwHdWEnjjOunhoQBs385B3p02Y4cVhKFFLMw91BY4 [2606:2800:220:1:248:1893:25c8:1946]: "<!doctype html>\n<html>\n<head>\n    <title>Example Domain</title>\n\n    <meta charset=\"utf-8\" />\n    <meta http-equiv=\"Content-type"
>
>  Domain: www.example.com
>  Type:   unauthorized
>  Detail: Invalid response from http://www.example.com/.well-known/acme-challenge/AJ1tjP3V02gi4X1r7qXu93m-BYL1qszD_TsQ7NNM2Fg [2606:2800:220:1:248:1893:25c8:1946]: "<!doctype html>\n<html>\n<head>\n    <title>Example Domain</title>\n\n    <meta charset=\"utf-8\" />\n    <meta http-equiv=\"Content-type"
>
>Hint: the Certificate Authority externally verifies the local changes that Certbot makes. Ensure the above domains are configured correctly and that changes made by the webroot plugin are accessible from the internet.

What I hope to achieve here (compared to having nothing) is to convey the basic mechanism by which automated domain validation works.

One thing that is bad here is that it's not exactly clear what the "changes" are, since it varies greatly by authenticator. If somebody got their `--webroot-path` wrong, would this message make the difference between them being able to self-troubleshoot or not? If not, is there any point including it?

Reply:
Oh my gosh I love this approach. I think the message at the top is super clear, and I love the framing of "Hint."

As for what to say in the hint, let's see. You're definitely right that it's not clear what "changes made by the webroot plugin" are. I do like the rest of the copy though, particularly the "accessible from the internet" phrase. (Just to check, is the idea that the text "webroot plugin" is generated by which plugin you're using? We definitely shouldn't be mentioning a plugin that they're not using.)

I wonder if we could mention `--debug-challenges`, and maybe expand on what we print there (instead of just saying "pass -v" which gets you way too much info). Like, we could print "challenges are located at path <generated path>" or "challenges are located in file x.conf". That might be a bit of work, depending on how the code is structured, though.

I mention putting it there because a) I think having the specific path would be nice rather than a general description of the sort of file/path, and once we get here it's too late to print the specific path since it won't exist anymore and b) I feel more comfortable printing slightly more in a place that's intentionally for someone who plans to do debugging themself, rather than asking for help.

Reply:
OK we decided on:

1. Per-authenticator "hints" if at all possible. 
  a) draft some copy
   b) try adapt `IAuthenticator`.
2. `--debug-challenges` is tricky because it doesn't work for the `renew` verb and changing that semantic was deemed too risky.

Here's my attempt at writing the hints for (1). I think it's pretty neat because we can make it responsive to whatever is in `IConfig` at the time of execution.  Potentially we could also make it responsive to the error code in `achalls`, but interpreting ACME problems is maybe not a good idea because IME even Boulder makes some questionable error code choices sometimes.

Some questions/comments are inline.

`--webroot`
-------
Hint: the Certificate Authority tried to download the the challenge files created by the --webroot plugin, but reported failure. Ensure that the provided --webroot-path/-w is correct for the above domains and that files created there are accessible from the internet.

- It would be nice to print a path directly here but `[webroot_map]` makes that kind of tricky. Keeping in mind, this hint is per-order and not per-challenge.


`--dns-*`
-------
Hint: the Certificate Authority tried to verify the DNS TXT records created by the --{} plugin, but reported failure. Ensure that the above domains are managed by that DNS provider, or try increasing --{}-propagation-seconds.


`--apache`
-------
Hint: the Certificate Authority tried to verify the temporary Apache configuration changes made by the --apache plugin, but reported failure. Ensure that the above domains point to this Apache server. 


`--nginx`
-------
Hint: the Certificate Authority tried to verify the temporary nginx configuration changes made by the --nginx plugin, but reported failure. Ensure that the above domains point to this nginx server, or try increase --nginx-sleep-seconds. 

- Do we want to document outstanding authenticator erratum/bugs in these hints? For example, --apache still has problems with duplicate ServerNames sometimes. --nginx is still broken if you try to use it when Cloudflare is being used in Full/Strict mode.


`--manual` (with hooks)
-------
Hint: the Certificate Authority tried to verify the changes made by the --manual-auth-hook, but reported failure. Ensure that this hook is functioning correctly{ and that it waits a sufficient duration of time after creating a DNS TXT record}. Refer to {certbot} --help manual.


`--manual` (no hook)
-------
Hint: the Certificate Authority tried to {verify/download} the manually created {DNS TXT records/challenge files}, but reported failure. Ensure that the {DNS TXT records/challenge files} were created in the correct location{ or try waiting longer for DNS propagation on the next attempt}.

- Possibly we want to add a hint in the "--manual pause notification" to wait for DNS propagation, instead of describing it later.

Reply:
- can we grab the flag/plugin names? if not, we'll need to make a note there to update the text here.
- I would rather not add text about outstanding authenticator bugs; maybe a pointer to the community forums if we know they're there? basically just don't want to have to update what will look like a random unrelated error message when we fix things

> Possibly we want to add a hint in the "--manual pause notification" to wait for DNS propagation, instead of describing it later.

I like this.

I also have some copy edits on this text, should those go inline?

Reply:
Approaching how to implement this, I don't think that `zope.interface.common.optional` class we came across works here. Putting an optional method into an interface sounds like nonsense to me :confused:. Shouldn't we create a new interface for that?

I'm thinking of putting a default `troubleshooting_hint()` into `certbot.plugins.common.Plugin` (or `ChallengePerformer` maybe), overriding it in relevant authenticators and not touching the zope interfaces. But I don't really have a great feel for what is going on here; what do you think?

Reply:
Yeah, that's better. That `optional` was too good to be true for sure. In practice, a quick look on github shows that most third party authenticators are for DNS plugins, either subclassing `dns_common.DNSAuthenticator`, `common.Plugin`, or using `--manual`, so if we have as much information as possible in the relevant places, that should cover those plugins, as well.

Reply:
Hmm, two more things:

1. We need a fallback hint for those authenticators that don't implement this. Rather than a vague idea of "changes", what about:
```python
        return ("The Certificate Authority couldn't exterally verify that the {name} plugin "
                "completed the required challenges. Ensure the plugin is configured correctly and "
                "that the changes it makes are accessible from the internet."
                .format(name=self.name))
```
2. Although we agreed not to try interpret the error state of the `achalls`, I realized that implementing the above text for the `--manual` plugin requires knowing what the attempted challenge type was. Do you think it's worth exposing the `achalls` to the hint function to enable this? Sacrificing the specificity would be a little disappointing :disappointed:.

Reply:
1. I like that text.
2. Ah, because manual doesn't always mean DNS. How bad would it be implementation-wise to expose the type? Would it help to just expose DNS vs. non-DNS? I think interpreting an error code and basing on the type of challenge requested are quite different, and this one is much more straightforward, with fewer cases, so I'm not as worried about this.

Reply:
Yeah, exposing the challenge type sounds good. Added that, albeit in a stringly way as I don't see an existing enum type for it.

One reason to instead expose the challs themselves, I think, would be for of future proofing. Say, if we wanted to expand the scope of this later on.

Reply:
*sadness noises*. I don't think we quite understood the challenges type hierarchy. Every challenge I get is descended from `challenges.KeyAuthorizationChallenge` (so either `challenges.DNS01` or `challenges.HTTP01`). I'm not sure when the `DNS` branch below is relevant:

https://github.com/certbot/certbot/blob/adb7e5e62f4c96be678c5af3e55a82ddcc717590/certbot/certbot/_internal/auth_handler.py#L283-L287

I'm going to try dig in this weekend to what `challenges.{KeyAuthorizationChallenge, _TokenChallenge, DNS, DNS01}` etc are and when they are used/instantiated, but if you know or know somebody who does, I'd appreciate that too.

Reply:
Documenting my own learning:

It looks `acme.challenges.DNS`  is from when `dns`, `simplehttp` and `dvsni` were around, in the original ACME draft (`draft-ietf-acme-acme-00`). I'd never read this version of the draft until today.

Then in #1271, `acme.challenges.KeyAuthorizationChallenge` was added as part of implementing `http-01`, which appeared in the `-01` draft. This was the first kind of challenge where the client produces a key authorization response, rather than signing the token using JWS.

I guess `dns` was kept around for backwards-compatibility reasons and it is indeed dead code.

So for this PR, I'm thinking that it's safe to assume that everything will be descended from `KeyAuthorizationChallenge`. 

I think that's all fine, but there's something we discussed yesterday that I forgot where we landed. I could feed in the list of failed `KeyAuthorizationAnnotatedChallenge`, or `acme.challenges.KeyAuthorizationChallenge`, or the `acme.messages.ChallengeBody`. Having to unwrap the challenge twice in `auth_hint` to get at the type information is a bit tedious. What do you think?

Reply:
Ahhhhh. 🤦😂😭 Well, nice find. That's tragic. Perhaps we should clean that up someday.

Are you sure it actually needs to be unwrapped? Don't these classes all have the sort of call-through where you can just ask the wrapping class for a property of the thing it wraps? Given that we're passing the whole object to try to keep information around, if there's information we lose by unwrapping I'd rather avoid that, but if we're not losing anything and we can't just ask the wrapper for the property, we can go ahead and unwrap before passing it in.

Reply:
>Don't these classes all have the sort of call-through

I'd seen it, but I just realized we can also use the call-through to avoid unwrapping the `challb`. That's nice.

That gives us a stringly-typed option:

```python
is_dns = any(achall.typ == challenges.DNS01.typ for achall in failed_achalls)
```

and a strongly-typed one:

```python
is_dns = any(isinstance(achall.chall, challenges.DNS01) for achall in failed_achalls)
```

I think that second one isn't too bad. Unless you know any Python tricks to make it more fluent, that checks all the boxes for me.

Reply:
The second one looks great. And that looks like good python to me, but Brad's the one who's into quirky python tricks if you wanted to find something else.

Reply:
>I also have some copy edits on this text, should those go inline?

Oops I missed this and the text in the diff has changed since then. Do you remember what these were?